MoltenVK/MoltenVK/Commands/MVKCommandBuffer.mm

```diff
@@ -334,7 +334,21 @@
 
     MTLRenderPassDescriptor* mtlRPDesc = [MTLRenderPassDescriptor renderPassDescriptor];
     getSubpass()->populateMTLRenderPassDescriptor(mtlRPDesc, _multiviewPassIndex, _framebuffer, _clearValues.contents(), _isRenderingEntireAttachment, loadOverride);
-    mtlRPDesc.visibilityResultBuffer = _occlusionQueryState.getVisibilityResultMTLBuffer();
+    if (_occlusionQueryState.getNeedsVisibilityResultMTLBuffer()) {
+        if (!_visibilityResultMTLBuffer) {
+            // Unfortunately, the temp buffer mechanism tends to allocate large buffers and return offsets into them.
+            // This won't work with visibility buffers, particularly if the offset is greater than the maximum supported
+            // by the device. So we can't use that.
+            // Use a local variable to make sure it gets copied.
+            id<MTLBuffer> visibilityResultMTLBuffer = [getMTLDevice() newBufferWithLength: _pDeviceMetalFeatures->maxQueryBufferSize options: MTLResourceStorageModePrivate];    // not retained
+            [visibilityResultMTLBuffer setPurgeableState: MTLPurgeableStateVolatile];
```

Comment:
Ugh! This could slow things down. `MTLBuffers` are really slow to allocate. And I've seen hundreds of Vulkan command buffers per frame. Not all doing occlusion of course...but still.

And the temp buffers also automatically handle the cleanup on command buffer completion, so they're just begging to be used here.

I'm not sure I understand the issue here. What's special about this use case? And what do you mean by an offset that can't be supported on the device?

Maybe (not necessarily on this PR) we can improve temp buffer allocation pooling to deal with whatever this use case needs.

Reply:
> I'm not sure I understand the issue here. What's special about this use case? And what do you mean by an offset that can't be supported on the device?

From doing tessellation, my experience with the temp buffer mechanism is that it tends to allocate large buffers, and then it carves regions out of them to use. Reading the source confirms that, for small buffers (i.e. 256k or smaller), this is indeed what it does. This is fine for most cases, but it doesn't work for visibility result buffers because the offset passed to -[MTLRenderCommandEncoder setVisibilityResultMode:offset:] cannot be greater than some implementation -defined maximum--formerly 64k less 8 bytes, for later GPUs 256k less 8 bytes. This is why query buffers are limited in size in the first place. If two buffers were taken from the temp buffer pool, the second buffer would have an offset exceeding this maximum, which could cause problems.

Reply:
>the offset passed to -[MTLRenderCommandEncoder setVisibilityResultMode:offset:] cannot be greater than some implementation-defined maximum

Ah. Thanks. Of course. Now I remember. That was the reason for the `supportLargeQueryPools` config setting.

Too bad `MTLRenderPassDescriptor setVisibilityResultBuffer:` doesn't allow for an offset to be set, as Metal often does elsewhere.

Okay. Let's go with this for now. I'll add to my own To-Do list to update the `MTLBuffer` pooling services to support the need for specifying a maximum offset (which might be zero). That will effectively dedicate a `MTLBuffer` per use, for this use-case, but at least they'll be pooled instead of transient.

I'll take care of that To-Do item myself on a separate PR.